sound/soc/sof/intel/hda-ipc.c

```diff
@@ -254,20 +252,62 @@ irqreturn_t hda_dsp_ipc_irq_handler(int irq, void *context)
 	if (irq_status == 0xffffffff)
 		goto out;
 
-	/* IPC message ? */
 	if (irq_status & HDA_DSP_ADSPIS_IPC) {
 		/* disable IPC interrupt */
 		snd_sof_dsp_update_bits_unlocked(sdev, HDA_DSP_BAR,
 						 HDA_DSP_REG_ADSPIC,
 						 HDA_DSP_ADSPIC_IPC, 0);
-		ret = IRQ_WAKE_THREAD;
+		ret = true;
 	}
 
 out:
-	spin_unlock(&sdev->hw_lock);
+	spin_unlock_irq(&sdev->hw_lock);
 	return ret;
 }
 
+/* is this IRQ for ADSP or SDW? - we only care about IPC and SDW here */
+irqreturn_t hda_dsp_ipc_irq_handler(int irq, void *context)
+{
+	struct snd_sof_dev *sdev = context;
+	struct sof_intel_hda_dev *hdev = sdev->pdata->hw_pdata;
+	int ret = IRQ_NONE;
+
+	sdev->irq_event = 0;
```

Comment:
we can't do this, the irq_event may be reset before it's used in a thread. I think we have no choice but to re-read the hardware registers in the thread.

Reply:
Thanks. @plbossart I didn't notice that. Maybe we can clear each bit in the irq thread the interrupt is handled.